src/main/scala/com/gwi/service/TaskServiceImpl.scala

```diff
@@ -0,0 +1,57 @@
+package com.gwi.service
+
+import akka.NotUsed
+import akka.actor.{ActorSystem, Cancellable}
+import akka.http.scaladsl.model.Uri
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import com.gwi.execution.{Task, TaskExecutor}
+import com.gwi.api.{TaskDetail, TaskState}
+import com.gwi.repository.TaskRepository
+import com.gwi.storage.TaskStorage
+
+import java.time.Instant
+import java.util.UUID
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ExecutionContext, Future}
+
+class TaskServiceImpl(taskRepository: TaskRepository, taskStorage: TaskStorage, taskExecutor: TaskExecutor)(implicit
+    ec: ExecutionContext,
+    system: ActorSystem
+) extends TaskService {
+
+  private val taskStateRefreshInterval = 2.seconds
+
+  override def createTask(csvUri: Uri): Future[UUID] = {
+    val newTask = Task(UUID.randomUUID(), csvUri)
+    for {
+      _ <- taskRepository.insertTask(newTask)
+      _ <- taskExecutor.enqueueTask(newTask.id)
+    } yield newTask.id
+  }
+
+  override def getTask(taskId: UUID): Future[Option[TaskDetail]] =
+    taskRepository.getTask(taskId).map(_.map(TaskDetail.fromTask))
+
+  override def getTaskSource(taskId: UUID): Source[Option[TaskDetail], Cancellable] =
+    Source
+      .tick(taskStateRefreshInterval, taskStateRefreshInterval, NotUsed)
+      .flatMapConcat(_ => Source.future(getTask(taskId)))
+      .takeWhile(_.exists(t => !TaskState.TerminalStates.contains(t.state)))
+
+  override def listTaskIds(): Future[List[UUID]] =
+    taskRepository.getTaskIds
+
+  override def cancelTask(taskId: UUID): Future[Either[String, UUID]] = {
+    taskRepository.getTask(taskId).flatMap {
+      case Some(task) if task.state == TaskState.Scheduled || task.state == TaskState.Running =>
+        taskExecutor.cancelTaskExecution(taskId)
```

Comment:
I would check if it was really canceled in order to not update tasks that are already `done`.